proto/private/utils/direct_source_infos.bzl

```diff
@@ -0,0 +1,72 @@
+# Copyright 2019 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Contains the implementation for `proto_common.direct_source_infos`."""
+
+ProtoFileInfo = provider(
+    doc = "Encapsulates information about a single `.proto` file.",
+    fields = {
+        "file": "The source file.",
+        "import_path": "The import path of the `.proto` file.",
+    },
+)
+
+# TODO(yannic): Remove this suppression when
+# https://github.com/bazelbuild/buildtools/issues/774 has been fixed.
+# buildifier: disable=function-docstring-args
+def direct_source_infos(proto_info, *, provided_sources = []):
+    """Returns sequence of `ProtoFileInfo` for `proto_info`'s direct sources.
+
+    Files that are both in `proto_info`'s direct sources and in
+    `provided_sources` are skipped. This is useful, e.g., for well-known
+    protos that are already provided by the Protobuf runtime.
+
+    Args:
+      proto_info: An instance of `ProtoInfo`.
+      provided_sources: Optional. A sequence of import paths to ignore.
+
+    Returns: A sequence of `ProtoFileInfo` containing information about
+        `proto_info`'s direct sources.
+    """
+
+    files = [_info(f, proto_info) for f in proto_info.direct_sources]
+    return [f for f in files if f.import_path not in provided_sources]
+
+def _info(file, proto_info):
+    """Computes `ProtoFileInfo` for a single file.
+
+    Args:
+      file: The proto file to generate `ProtoFileInfo` for. Must be in
+          `direct_sources` of `proto_info`.
+      proto_info: An instance of `ProtoInfo`.
+
+    Returns: An instance of `ProtoFileInfo`.
+    """
+
+    if "." == proto_info.proto_source_root:
+        # The `proto_library` didn't specify `import_prefix` or
+        # `strip_import_prefix`, and `file` is a regular source file
+        # (i.e. not generated) in the main workspace.
+
+        return ProtoFileInfo(
+            file = file,
+            import_path = file.path,
+        )
+
+    source_root = proto_info.proto_source_root
+    offset = len(source_root) + 1  # + '/'.
```

Comment:
Whoa, I think I caught an *actual* bug. This logic seems to want to do same thing as the the exec path -> import path logic from the Java code:

https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/rules/proto/ProtoCompileActionBuilder.java#L633

but it doesn't do that, AFAIU: it diverges for `.proto` files in external repositories and generated `.proto` files (yes, they do exist)

Do I understand this correctly? If so, it might be more reasonable to expose this information from Java in the interest of not re-implementing the same functionality twice. WDYT?

Reply:
I'm not sure I understand the case in which it doesn't work.

We do have coverage for external repositories, so I'm confident that works correctly. We don't have tests for generated `.proto` files here, but aren't they always simlink'ed into `<pkg>/_virtual_imports/<name>/<import_path>` since https://github.com/bazelbuild/bazel/issues/9215, thus work as well?

The only case I see when this logic could fail is when the file is not under `proto_source_root` (which, AFAIK, cannot happen since Bazel 1.0). What am I missing?

Reply:
You're totally right! However, there is the problem of the direct proto source root being a lie (e.g. if a `proto_library` has both source and derived `.proto` files, not all of them are going to be under `.proto_source_root`.

I suppose https://github.com/bazelbuild/bazel/pull/10939 will help with that?